[FLINK-15102] Add continuousSource() method to StreamExecutionEnvironment.

becketqin opened the pull request:
## What is the purpose of the change
This patch is a part of FLIP-27. It adds `continuousSource()` set of methods to the `StreamExecutionEnvironment`. The methods take a `Source` and returns a `DataStream`.

## Brief change log
- Added the `continuousSource(Source)` set of methods to the `StreamExecutionEnvironment`.
- Added unit tests.
- Added `NoMoreSplitsEvent` to the base source implementation so a bounded source can exit eventually.

## Verifying this change
This change added tests and can be verified as follows: 
* org.apache.flink.streaming.api.graph.StreamingJobGraphGeneratorTests.testCoordinatedOperator
* org.apache.flink.connector.base.source.reader.CoordinatedSourceITCase

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

TsReaper commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:
```diff
@@ -432,6 +434,20 @@ private StreamConfig createJobVertex(
 					chainedNames.get(streamNodeId),
 					jobVertexId,
 					operatorIDPairs);
+			if (streamNode.getOperatorFactory() instanceof CoordinatedOperatorFactory) {
+				OperatorCoordinator.Provider coordinatorProvider =
+						((CoordinatedOperatorFactory<?>) streamNode
+								.getOperatorFactory())
+								.getCoordinatorProvider(
+										chainedNames.get(streamNodeId),
+										new OperatorID(hashes.get(streamNodeId)));
+				try {
+					jobVertex.addOperatorCoordinator(new SerializedValue<>(coordinatorProvider));
+				} catch (IOException e) {
+					throw new FlinkRuntimeException(String.format(
+							"Coordinator Provider for node %s is not serializable.", chainedNames.get(streamNodeId)));
+				}
+			}
```
This will only add the coordinator of the head of the operator chain into the job vertex, coordinators of other operators will be lost.

StephanEwen replied:
I think @TsReaper's comment here is correct.

The proposed solution is a bit hacky, but a clean solution would need some major prerequisite cleanup of the whole StreamGraph, and StreamingJobGraphGenerator, which would be a separate bigger effort.

StephanEwen commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java:
```diff
@@ -81,6 +82,8 @@ Licensed to the Apache Software Foundation (ASF) under one
 	/** The last element to ensure it is fully handled. */
 	private SplitsRecordIterator<E> splitIter;
 
+	private volatile boolean noMoreSplitsAssignment;
```
Does this need to be volatile? Isn't it updated always by the same thread (mailbox thread)?

StephanEwen commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java:
```diff
@@ -266,7 +268,10 @@ public boolean isIterative() {
 
 		if (operatorFactory.isStreamSource()) {
-			addNode(vertexID, slotSharingGroup, coLocationGroup, SourceStreamTask.class, operatorFactory, operatorName);
+			Class<? extends AbstractInvokable> invokableClass =
```
This looks fragile to me. See the top-level comment about an idea for a cleaner way of handling this.

becketqin commented:
@StephanEwen Thanks for the review and suggestions. While I was addressing the comments, I realized that a class of `CoordinatedOperatorTransformation` may not be the best approach because this disallows the users to have coordinators in other transformations. So instead, I added a method of `getOperatorCoordinator()` method to the `PhysicalTransformation`.  The method returns an `Optional` so the transformations can override it if necessary.

I also adjusted `StreamingJobGraphGenerator#createChain()` a little bit so the logic looks slightly clearer.

Just curious about the convention here. So is the convention only calling `instanceOf` in the `StreamGraphGenerator` and have typed parameter everywhere else?

@TsReaper Thanks for the review. The latest update should have fixed the problem. Would you like to take another look?

TsReaper commented:
Thanks for the update @becketqin , the registration of coordinators into job vertex looks good to me now.

StephanEwen commented:
The update looks fine to me.

It would be nice to address the comment about the `CoordinatedSourceITCase` (more thorough result checking), otherwise +1 to merge this

becketqin commented:
Thanks @StephanEwen. Apologies that I somehow missed some of your previous comments...

I'll merge the patch with the following changes.

1. `SourceReader` actually does not have to extend `Serializable`. So I just removed that from the interface. I also fixed other places as suggested where there is a warning.
2. Change the `CoordinatedSourceITCase` to compare the full list.

StephanEwen commented:
Thanks for addressing the comments. Looks good to me, @becketqin.

+1 to merge this

TsReaper commented:
Sorry for interrupting... But I discover that in the current code we still can only register coordinators for source. How can I register a coordinator for a sink?

becketqin commented:
@TsReaper I am fixing the compilation error. It was because a recent commit changed the interface of SourceOperator.

From what I understand, all the operators will be provided through factories in the future. So does the sink. In that we may have to introduce a new method in the `DataStream` API in order to let the Sink leverage this mechanism. In terms of the `Table.collect()`. I think some modifications need to be made to `StreamGraphGenerator.transformSink()` and `StreamGraph.addSink()` to pass in the coordinator provider.

becketqin commented:
@StephanEwen Sorry for the back and forth, after second thought, I feel it might makes sense to put the `instanceOf CoordinatedOperatorFactory` in `StreamNode` instead of in the `PhysicalTransformation`. The reason being is following:

0. `CoordinatedOperatorFactory` is an implementation detail of a `StreamOperatorFactory`, instead of a type of it. It is orthogonal to and can be implemented by any type of other `StreamOperatorFactory`.
1. Conceptually speaking, `PhysicalTransformation` only cares about the type of the operator factory, it does not care about the implementation, i.e. whether the operator has a coordinator. Therefore the only thing `PhysicalTransformation` can do is to extract the coordinator provider and pass it to the `StreamGraphGenerator`. It seems much simpler and more intuitive to just let `StreamNode` to do the `instanceOf` and extract the coordinator provider factory in a method like 
```
Optional<OperatorCoordinator.Provider> getOperatorCoordinatorProvider(String operatorName, OperatorID operatorID)
```
2. Another important benefit is that this makes sure we only have one single place that needs to care about `CoordinatedProviderFactory`, which is much less error prone.

The last commit updated the patch with this change. What do you think? 

If you have concern for it, we can merge the PR without the last commit and make the change later if we would like to.

StephanEwen commented:
Your argument makes sense.

I am also not opinionated about the `StreamGraph`, `StreamGraphGenerator` and `StreamingJobGraphGenerator` - these classes are quite messy and need a wholesome cleanup anyways, so in the end you should go with whatever is the simplest solution here.

becketqin commented:
Merged to master.
7abced6d7253845270e555f044fd2db774a54e3d
